File: binary_search.java
// Given an array of integers nums which is sorted in ascending order, and an integer target, write a function to search target in nums. If target exists, then return its index. Otherwise, return -1.
/* Binary search is a basic method that compares the target value to the array's middle member.

If the goal value equals the center element, we're done.



If the goal value is lower, continue to search on the left.



If the goal value is higher, continue searching on the right. */


class Solution {
  public int search(int[] nums, int target) {
    int pivot, left = 0, right = nums.length - 1;
    while (left <= right) {
      pivot = left + (right - left) / 2;
      if (nums[pivot] == target) return pivot;
      if (target < nums[pivot]) right = pivot - 1;
      else left = pivot + 1;
    }
    return -1;
  }
}

File: binarySearch.py
"""
Binary search is a basic method that compares the target value to the array's middle member.

If the goal value equals the center element, we're done.



If the goal value is lower, continue to search on the left.



If the goal value is higher, continue searching on the right.

"""


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1
        while left <= right:
            pivot = left + (right - left) // 2
            if nums[pivot] == target:
                return pivot
            if target < nums[pivot]:
                right = pivot - 1
            else:
                left = pivot + 1
        return -1
